panel/panel.py

```diff
@@ -207,107 +175,103 @@ def fill_panel(window, update=False, **panel_filter):
+            base_lineno = len(to_render)
+            for i, item in enumerate(buf_lines):
+                to_render.append(format_row(item))
+                item["panel_line"] = base_lineno + i
 
             # insert empty line between views sections
             to_render.append("")
 
-    rendered_text = to_render.render() if to_render else None
-    run_update_panel_cmd(panel, text=rendered_text)
+    run_update_panel_cmd(panel, text="\n".join(to_render))
 
 
 # logic for updating panel selection
 
-def get_closest_region_dict(dic, colno):
-    dics = [
-        d
-        for error_dict in dic.values() for d in error_dict
-        if d.get("panel_lineno")
-        # if d["start"] <= colno <= d["end"]  # problematic line
-    ]
-    if not dics:
-        return
-    return min(dics, key=lambda x: abs(x["start"] - colno))
-
+def get_next_panel_line(buf_line, buf_dict):
+    """Get panel line for next error in buffer.
 
-def get_next_lineno(num, interval):
-    interval = set(interval)
-    interval.discard(num)
-    interval = list(interval)
-    interval.sort()
-
-    if num < interval[0] or interval[-1] < num:
-        return interval[0]
+    If not found this means the current line is past last error's buffer line.
+    In that case return last error's panel line incremented by one, which will
+    place panel selection in empty space between buffer sections.
+    """
+    for d in buf_dict:
+        if d["line"] > buf_line:
+            panel_line = d["panel_line"]
+            break
     else:
-        i = bisect.bisect_right(interval, num)
-        neighbours = interval[i - 1:i + 1]
-        return neighbours[1]
+        panel_line = buf_dict[-1]["panel_line"] + 1
 
+    return panel_line, panel_line
 
-def change_selection(panel_lineno, full_line=False):
-    panel = get_panel(sublime.active_window())
-    if not panel:
-        return
-
-    region = sublime.Region(panel.text_point(panel_lineno - 1, -1))
-    if full_line:
-        region = panel.line(region)
 
+def update_selection(panel, region=None):
     selection = panel.sel()
     selection.clear()
-    selection.add(region)
+    if region is not None:
+        selection.add(region)
 
-    # scroll selection into panel
-    if not panel.visible_region().contains(region):
-        panel.show(region)
 
-    # simulate scrolling to enforce rerendering of panel,
-    # otherwise selection is not updated (ST core bug)
-    panel.run_command("scroll_lines")
+def get_panel_region(row, panel, is_full_line=False):
+    region = sublime.Region(panel.text_point(row - 1, -1))
+    if is_full_line:
+        region = panel.line(region)
+    return region
 
 
 def update_panel_selection(active_view, we_count, current_pos, **kwargs):
     if current_pos == (-1, -1):
         return
 
-    full_line = False
-    view_dict = persist.errors.get_view_dict(active_view.id())
-    if not view_dict or util.is_none_or_zero(we_count):
+    buf_dicts = persist.raw_errors.get(active_view.buffer_id())
```

Comment:
I have noticed a `NoneType exception` in the console once. How could this be with a defaultdict (`{}`).
Shouldn't the default value be an empty dict instead?
Wished that guard immediately below could go.

Reply:
Using `get` you get None's bc otherwise you would add keys to the dict when you would like to check if a key is in the dict. defaultdict replaces exactly what we do with `setdefault(key, {})`

`bar.setdefault('foo', []).append('fox') -- becomes --> bar['foo'].append('fox')`

It is more readable and faster.